stdlib/asm/crypto/fri/frie2f4.masm

```diff
@@ -270,28 +338,152 @@ export.verify.4
+        swapw
+        # => [x, x, x, x, query_ptr + 1, layer_ptr, rem_ptr, g, ...]
+        dup.5
+        dup.5
         neq
-        #=> [?, query_ptr + 4, layer_ptr, rem_ptr, g, ...]
+        #=> [?, query_ptr + 1, layer_ptr, rem_ptr, g, ...]
     end
     #=> [X, ..]
 
-    dropw
+    dropw dropw
+end
+
+#! Verifies a FRI proof where the proof was generated over the quadratic extension of the base
+#! field and layer folding was performed using folding factor 4 when the degree of the remainder
+#! polynomial is less than 128.
+#!
+#! Input:  [query_ptr, layer_ptr, rem_ptr, g, ...]
+#! Output: [...]
+#!
+#! - query_ptr is a pointer to a list of tuples of the form (e0, e1, p, poe) where poe is equal
+#!   to g^p with g being the initial FRI domain generator. p is the query index at the first layer
+#!   and (e0, e1) is an extension field element corresponding to the value of the first layer at index p.
+#! - layer_ptr is a pointer to the first layer commitment denoted throughout the code by C.
+#!   layer_ptr + 1 points to the first [alpha0, alpha1, t_depth, d_size] where d_size is the size
+#!   of initial domain divided by 4, t_depth is the depth of the Merkle tree commitment to the
+#!   first layer and (alpha0, alpha1) is the first challenge used in folding the first layer.
+#!   Both t_depth and d_size are expected to be smaller than 2^32. Otherwise, the result of
+#!   this procedure is undefined.
+#! - rem_ptr is a pointer to the first tuple of two consecutive degree 2 extension field
+#!   elements making up the remainder polynomial. This procedure is specialized to the case when
+#!   the the degree of the latter is less than 128.
+#!
+#! The memory referenced above is used contiguously, as follows:
+#!
+#!   [query_ptr ... layer_ptr ... rem_ptr ...]
+#!
+#! This means for example that:
+#! 1. rem_ptr - 1 points to the last (alpha0, alpha1, t_depth, d_size) tuple.
+#! 2. layer_ptr - 1 points to the last (e0, e1, p, poe) tuple.
+#!
+#! Cycles: 24 + num_queries * (140 + num_layers * 83)
+proc.verify_128.4
+
+    # store [query_ptr, layer_ptr, rem_ptr, g] to keep track of all queries
+    # (3 cycles)
+    loc_storew.0
+
+    # [(query_ptr == layer_ptr), query_ptr, layer_ptr, rem_ptr, g]
+    # (4 cycles)
+    dup
+    dup.2
+    neq
+
+    # Save a word containing a fresh accumulator for Horner evaluating the remainder polynomial,
+    # a pointer to the evaluation point and a pointer to the location of the polynomial.
+    push.0.0 
+    exec.constants::tmp2 exec.constants::get_remainder_poly_address
+    exec.constants::tmp1 mem_storew
+    movup.4
+    
+    while.true
+        # load [e0, e1, p, poe] from memory i.e. next query data (7 cycles)
+        movup.4
+        mem_loadw
+        # => [poe, p, e1, e0, layer_ptr, rem_ptr, g, ...]
+
+        # we now have everything to verify query p
+        exec.verify_query_128
+
+        # prepare for next iteration (18 cycles)
+        # => [x, x, x, x, x, x, x, x, x, x, g, ...]
```

Comment:
Similarly here, there seem to be 12 garbage elements on the stack at the end of `verify_query_128`, but this comment suggests there are 10

Reply:
fixed